### PR TITLE
Alerting: Add size checks on text messages for the Telegram notifier

### LINKED ISSUE #83412: Improve Telegram channel's performance with large alert notifications
### Discussed in https://github.com/grafana/grafana/discussions/48974

<div type='discussions-op-text'>

<sup>Originally posted by **pcolladosoto** May  6, 2022</sup>
## What happened?
We've configured a contact point serving alerts both through *email* and *Telegram* on our Grafana instance. We are currently under a heavy alert load and we have noticed notifications are flawlessly delivered through email messages but the same is not true for Telegram...

## What you expected to happen
We expected the notifications to be correctly delivered **both** over email and Telegram instead on just through the former.

## How to reproduce it
TL;DR: Make a fairly large number of alerts (35 in our case) trigger to prevent Telegram from sending notifications.

We observed Telegram stopped sending alerts when under pressure in terms of the number of alerts. In order to test stuff out we prepared a collection of 'synthetic' alerts we could trigger at will. We found that when all the alerts were triggering Telegram wasn't capable of sending them. On the other hand, when we silenced most of them through the use of *tags* we suddenly discovered Telegram was sending the alerts again!

## Anything else we need to know?
Given the behaviour we've just explained above we decided to take a close look at Grafana's logs to try and shed some light on the issue. In order to do so we added the following option to increase the log's verbosity on the config file (i.e. `/etc/grafana/grafana.ini`):

```ini
[log]
level = debug
```

After doing so we inspected our logs (over at `/var/log/grafana/grafana.log`) and we found the following:

```
t=2022-05-06T13:19:13+0200 lvl=info msg="sending telegram notification" logger=alerting.notifier.telegram chat_id=MyChannelId
t=2022-05-06T13:19:13+0200 lvl=dbug msg="Sending webhook" logger=notifications url=https://api.telegram.org/botMyBotToken/sendMessage http method=POST
t=2022-05-06T13:19:13+0200 lvl=dbug msg="Webhook failed" logger=notifications url=https://api.telegram.org/botMyBotToken/sendMessage statuscode="400 Bad Request" body="{\"ok\":false,\"error_code\":400,\"description\":\"Bad Request: message is too long\"}"
t=2022-05-06T13:19:13+0200 lvl=eror msg="Failed to send webhook" logger=alerting.notifier.telegram error="Webhook response status 400 Bad Request" webhook=Default
t=2022-05-06T13:19:13+0200 lvl=info msg="level=error component=dispatcher msg=\"Notify for alerts failed\" num_alerts=35 err=\"Default/telegram[0]: notify retry canceled due to unrecoverable error after 1 attempts: Webhook response status 400 Bad Request\"" logger=alertmanager org=1
```

They key bit is the one that says:

```
"{\"ok\":false,\"error_code\":400,\"description\":\"Bad Request: message is too long\"}"
```

Given we are invoking the [`sendMessage`](https://core.telegram.org/method/messages.sendMessage#return-errors) method on Telegram's API we can just check the error codes there. However, given the `description` we can be sure about what the likely cause of the problem is: we are trying to fit in way too many characters in the messages body. Given we're trying to squeeze in a ton of alerts that's something that can make sense.

When we have a couple of alerts triggering notifications make it through and we see the following entries in the log:

```
t=2022-05-06T13:24:13+0200 lvl=info msg="sending telegram notification" logger=alerting.notifier.telegram chat_id=MyChannelId
t=2022-05-06T13:24:13+0200 lvl=dbug msg="Sending webhook" logger=notifications url=https://api.telegram.org/botMyBotToken/sendMessage http method=POST
t=2022-05-06T13:24:13+0200 lvl=dbug msg="Webhook succeeded" logger=notifications url=https://api.telegram.org/botMyBotToken/sendMessage statuscode="200 OK"
t=2022-05-06T13:24:13+0200 lvl=info msg="level=debug component=dispatcher receiver=Default integration=telegram[0] msg=\"Notify success\" attempts=1" logger=alertmanager org=1
```

Another key behaviour is that Telegram's Test notifications always seem to work... unless we make them too big! In order to do so we can just write a ton of bogus data on the *Message* field and we will trigger the same error...

So, all in all, we have experienced how Telegram won't notify us of alerts when we have a fairly large number of them triggering...

We don't know if this is the desired behaviour when managing these notifications, but we have been very thrown away by not being notified.

If this were something to alter we believe we could send a message per notification so as not to cross the message length threshold or, at least, surpass it less often. This change in the way of delivering notifications could maybe trigger rate limits of bots when sending messages. We could maybe circumvent this possible hindrance with some sort of rate limiting should we face issues of that nature.

In any case, we're more than happy to hear your suggestions. Note that even though we flagged this as a bug that might not be the case: feel free to change the issue's category as you deem appropriate.

Thanks for your time and sorry for the 'noise'! 😸

## Environment
- Grafana version: 8.3.4
- Data source type & version: InfluxDB v1.8.0
- OS Grafana is installed on: CentOS 7
- User OS & Browser: Safari on macOS.
- Grafana plugins: Just the core plugins.
- Others: Nothing we can think of 😅</div>

### PR DESCRIPTION
**Pull request by pcolladosoto**
**What is this feature?**

As described on issue #83412, sending text-based Telegram notifications won't work when the message body becomes too large. We believe this was caused by the lack of size checks when constructing the message to be sent when leveraging the `sendMessage` API endpoint instead of `sendPhoto`. This PR includes size checks for the former case so that alerts do get sent even if they're a bit too large.

**Why do we need this feature?**

As seen on issue #83412, if the resulting alert message when using `sendMessage` (i.e. when calling `buildMessageLinkedImage`) exceeded Telegram's maximum message size it wouldn't be sent at all. This would blindside the recipients of the alerts, as it would seem as if no alerts where being generated...

Even if the alert message that's sent is not complete, the reception of a message will alert admins so that they can investigate further and inquire so as to what's triggering the alerts. In other words, the notification will serve the purpose of, well, notifying people in charge of reacting to abnormal conditions.

**Who is this feature for?**

Everyone using Telegram as an alerting channel in their Grafana instances.

**Which issue(s) does this PR fix?**:

Fixes #83412

**Special notes for your reviewer:**
According to https://limits.tginfo.me/en, the message size limit sits at 4096 UTF-8 characters, so we have enforced that limit when constructing the message to be sent out. What's more, we have also added tests to check this condition is met when the message payload is too large.

The strategy to shorten the message is the same as the one followed by existing logic related to the `sendPhoto` flow: the message will be truncated to the maximum size and any data that's appended to it must fit within the known limit.

Tests have been updated to exercise the new implementation in a fashion that's backwards-compatible with existing methods.

Should any issues exist or should any improvements be carried out we would be delighted to work on it. Thanks a ton for your time and for making Grafana!

### CODE REVIEW

**Review by yuri-tceretian — COMMENTED** [review]
@pcolladosoto thanks for contributing! However, this telegram integration is the legacy one and it's going to be deleted very soon. 
Please refer to https://github.com/grafana/alerting/blob/ec2c02267fa591e7252ddb44143555bc941ce6fa/receivers/telegram/telegram.go in the `grafana/alerting` repository

**Comment by pcolladosoto** [comment]
Hi @yuri-tceretian! Thanks a ton for the update: I had no idea... I looked at the new implementation and it looks great 😼. Thanks for taking the time to look over the code 😸

**Comment by fopina** [comment]
@yuri-tceretian I see the "new notification service" mentioned in quite a few issues but I don't understand how I can use it at the moment.
Is it available in Grafana 10? If I'm choosing Telegram as contact point, how do I make sure it's using ng instead of legacy?

Really need these messages split 🙏

**Comment by fopina** [comment]
Thanks @yuri-tceretian, confirmed!

**Comment by fopina** [comment]
Actually, I continue to get failed alerts (telegram error 400) even for truncated messages with ngalert.

I assume it's likely due to html format and emojis being used, I'll try to setup reproducible steps

**Comment by feym78** [comment]
Still having this issue.
Using Grafana 10.1.5 and custom template for telegram contact point we had a bunch of alerts (about 36) and all alerts stopped. Interesting thing that when we resolved alerts - resolve message came truncated, but firing and resolve templates in general are same.

**Comment by fopina** [comment]
@feym78 might be the same issue I've described in https://github.com/grafana/alerting/issues/185
Some get truncated mid-html/markdown and get error 400, others get truncated without breaking html and are sent

**Comment by feym78** [comment]
> @feym78 might be the same issue I've described in [grafana/alerting#185](https://github.com/grafana/alerting/issues/185) Some get truncated mid-html/markdown and get error 400, others get truncated without breaking html and are sent

thanks a lot, looks like indeed it's the same issue

**Comment by yuri-tceretian** [comment]
The code this PR updates was deleted in Grafana 11. I would appreciate it if you could open a fix for the current version of Telegram integration https://github.com/grafana/alerting/tree/main/receivers/telegram. Please tag me in a PR and I promise I will review it asap.

**Comment by pcolladosoto** [comment]
Hi @yuri-tceretian! It looks like this issue has already been resolved on https://github.com/grafana/alerting/commit/a5c9ae1d99fb02739cbd6a48d1fd9129cd4bb5e

Thanks a ton for the update!
